tests/unit/s2n_server_psk_extension_test.c

```diff
@@ -0,0 +1,514 @@
+    GUARD_AS_RESULT(s2n_psk_init(other_server_psk, S2N_PSK_TYPE_EXTERNAL));
+    GUARD_AS_RESULT(s2n_psk_new_identity(other_server_psk, other_server_data, sizeof(other_server_data)));
+    GUARD_AS_RESULT(s2n_psk_new_secret(other_server_psk, other_server_data, sizeof(other_server_data)));
+    other_server_psk->hmac_alg = S2N_HMAC_SHA224;
+
+    return S2N_RESULT_OK;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Test: s2n_server_psk_should_send */
+    {
+        struct s2n_psk *psk = NULL;
+
+        struct s2n_connection *conn = NULL;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_SERVER));
+
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(NULL));
+
+        conn->actual_protocol_version = S2N_TLS12;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+        conn->actual_protocol_version = S2N_TLS13;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+
+        EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &psk));
+
+        conn->actual_protocol_version = S2N_TLS12;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+        conn->actual_protocol_version = S2N_TLS13;
+        EXPECT_FALSE(s2n_server_psk_extension.should_send(conn));
+
+        conn->psk_params.chosen_psk_wire_index = 0;
+        EXPECT_OK(s2n_array_get(&conn->psk_params.psk_list, conn->psk_params.chosen_psk_wire_index,
+                                (void **)&conn->psk_params.chosen_psk));
+        EXPECT_TRUE(s2n_server_psk_extension.should_send(conn));
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test: s2n_server_psk_send */
+    {
+        /* Send the index of the chosen PSK that is stored on the connection. */
+        {
+            struct s2n_stuffer out = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+
+            struct s2n_connection *server_conn = NULL;
+            struct s2n_connection *client_conn = NULL;
+            EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+            EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+            struct s2n_psk *shared_psk = NULL;
+            EXPECT_OK(setup_order_mismatch_psks(server_conn, client_conn, &shared_psk));
+            EXPECT_NOT_NULL(shared_psk);
+
+            uint16_t shared_psk_wire_index = 0;
+
+            for (size_t i = 0; i < client_conn->psk_params.psk_list.len; i++) {
+                struct s2n_psk *psk = NULL;
+                EXPECT_OK(s2n_array_get(&client_conn->psk_params.psk_list, i, (void**)&psk));
+                if (psk == shared_psk) {
+                    shared_psk_wire_index = i;
+                }
+            }
+
+            /* Valid chosen psk wire index set */
+            server_conn->psk_params.chosen_psk_wire_index = shared_psk_wire_index;
+            EXPECT_SUCCESS(s2n_server_psk_extension.send(server_conn, &out));
+
+            uint16_t chosen_psk_wire_index;
```

Comment:
Initializing this value to 0 has a different meaning as it would mean that the chosen_psk_wire_index is the first index in the wire list.  The correct initialization should be the value sent by s2n_server_psk_extension.send and is initialized in the line below.